File: 跟踪滤波设计包/matlab/validate_residual_alignment.m
function validate_residual_alignment()
%VALIDATE_RESIDUAL_ALIGNMENT Review item #4: residual_mode == gear_filter.
%   MATLAB/Octave port of validate_residual_alignment.py: end-to-end
%   consistency of the product path (residual_mode) vs the validation path
%   (vt_gear_filter) over 3 gears x 3 frequencies x {near-noiseless/always,
%   CNR=3dB/auto}; asserts |err_gear - err_core| < 1 percentage point.
%
%   Run:  cd matlab && octave --eval validate_residual_alignment
%   Saves golden metrics to golden/validate_residual_alignment_mat.mat and
%   raises an error (nonzero exit code) if any combination fails.
  t_all = tic;
  sd = fileparts(mfilename('fullpath'));
  addpath(fullfile(sd, 'homodyne'));
  ensure_kernels();

  dp = design_params();
  c = vt_const();
  TOL_PP = 1.0;
  FREQS = [100e3, 1e6, 3e6];

  fprintf(['残差窗一致性验证 (审查项 #4): core.residual_mode vs ' ...
           'validate_tracking.gear_filter\n']);
  fprintf('  fs=%.0fMS/s, B_win=%.0fMHz, NT_WIN=%d taps, 容差 |Δerr| < %.0fpp\n', ...
          dp.FS / 1e6, dp.B_WIN / 1e6, dp.NT_WIN, TOL_PP);

  tags = {'near-noiseless / gate=always', 'CNR=3dB noisy / gate=auto'};
  gates = {'always', 'auto'};
  cnr_dbs = [NaN, 3.0];
  n_cases = 2 * 3 * 3;
  err_gear = zeros(n_cases, 1);
  err_core = zeros(n_cases, 1);
  adiff = zeros(n_cases, 1);
  okv = false(n_cases, 1);
  idx = 0;
  nfail = 0;
  for it = 1:2
    gate = gates{it};
    fprintf('\n  [%s]\n', tags{it});
    fprintf('    %-7s %7s | %15s %17s | %9s\n', 'gear', 'f0', ...
            'err gear_filter', 'err residual_mode', '|diff| pp');
    for ib = 1:3
      band = dp.ORDER{ib};
      for ifq = 1:3
        f0 = FREQS(ifq);
        sc = vt_make_scene(f0);
        if strcmp(gate, 'always')
          z = vt_clean_z(sc);
          Nhat = 1e-10;
        else
          s2 = 10^(-cnr_dbs(it) / 10);
          rh = np_rng_new(40000 + fix(f0 / 1e3));
          z = exp(1i * sc.ph) + ...
              complex_bandlimited_noise(c.N, dp.FS, dp.B_FRONTEND, s2, rh);
          Nhat = s2;
        end
        yg = vt_gear_filter(z, band, Nhat, gate);
        ropts = gate_params(band);
        ropts.zeta = dp.ZETA;
        ropts.tauG = dp.TAU_G;
        ropts.Nt_win = dp.NT_WIN;
        ropts.gate = gate;
        yc = residual_mode(z, dp.FS, dp.BANDS.(band).fn, Nhat, ...
                           dp.B_WIN, ropts);
        eg = vt_amp_err_pct(vt_vdisc(yg), sc);
        ec = vt_amp_err_pct(vt_vdisc(yc), sc);
        d = abs(eg - ec);
        ok = d < TOL_PP;
        idx = idx + 1;
        err_gear(idx) = eg;
        err_core(idx) = ec;
        adiff(idx) = d;
        okv(idx) = ok;
        if ok, mark = ''; else, mark = '   <-- FAIL'; nfail = nfail + 1; end
        fprintf('    %-7s %5.0fk | %+14.3f%% %+16.3f%% | %9.4f%s\n', ...
                band, f0 / 1e3, eg, ec, d, mark);
      end
    end
  end

  fprintf('\n%s\n', repmat('=', 1, 70));
  if nfail > 0
    fprintf('FAIL: %d/%d 组合差异 >= %.0fpp\n', nfail, n_cases, TOL_PP);
  else
    fprintf(['PASS: 全部 %d 组合 (三档 x 三频 x 无噪/含噪) 幅度误差差异 < ' ...
             '%.0fpp -- 验证路径 = 产品路径\n'], n_cases, TOL_PP);
  end
  fprintf('[elapsed %.1f s]\n', toc(t_all));

  % ------------------------------------------------- golden metrics (.mat)
  g = struct();
  g.checks_ok = double(okv(:)');
  g.checks_pass = sum(g.checks_ok);
  g.checks_total = n_cases;
  g.noisy = struct('err_gear', err_gear, 'err_core', err_core, ...
                   'adiff', adiff);
  gdir = fullfile(sd, 'golden');
  if ~exist(gdir, 'dir'), mkdir(gdir); end
  gfile = fullfile(gdir, 'validate_residual_alignment_mat.mat');
  save('-v7', gfile, '-struct', 'g');
  fprintf('[golden metrics saved to %s]\n', gfile);

  if nfail > 0
    error('validate_residual_alignment: SOME CHECKS FAILED');
  end
end

File: 跟踪滤波设计包/homodyne_tracking_design/validate_residual_alignment.py
#!/usr/bin/env python3
"""审查项 #4: core.residual_mode (产品路径) vs gear_filter (验证路径) 一致性断言.

背景: 原 core.residual_mode 的残差测量窗是一阶 IIR (iir1_lowpass), 而
validate_tracking.py 验证的是 1025-tap 线性相位 FIR 窗 -- 验证 PASS 不覆盖
产品路径, 且 residual_mode 是死代码 (无人调用).  现 residual_mode 已改为与
gear_filter 复用同一 FIR 设计函数 (core.fir_lp_kernel, Hann 窗加窗 sinc,
NT_WIN=1025 taps), 本脚本对两条路径做端到端一致性断言:

  SLOW/MEDIUM/FAST 三档 x 100 kHz / 1 MHz / 3 MHz 三频, 分别在
    (a) 近无噪 burst, gate='always'  (validate_tracking 规则 R1 的幅值转移)
    (b) 含噪 CNR=3 dB, gate='auto'   (同一噪声实现喂给两条路径)
  比较 lock-in 幅度误差 (%), 断言 |err_gear - err_core| < 1 个百分点.

群时延说明: 两条路径均以 'same' 对齐补偿 (Nt-1)/2 采样 FIR 群时延;
实时硬件需在 NCO 相位路径加 NT_WIN/2 采样延迟线 (见 core.fir_lp_same 注释).
250 MS/s 全速率单级 1025 taps 硬件不可行, 需多级降采样等效实现
(见 design_params.NT_WIN 注释); 本脚本与 validate_tracking 一样直接跑
全速率参考滤波器.
"""
import time
import numpy as np

from core import residual_mode, complex_bandlimited_noise
from design_params import (
  FS, B_FRONTEND, ZETA, B_WIN, NT_WIN, TAU_G, BANDS, ORDER, gate_params,
)
from validate_tracking import (
  N, gear_filter, make_scene, clean_z, amp_err_pct, vdisc,
)

TOL_PP = 1.0                     # 断言容差: 幅度误差差异 < 1 个百分点
FREQS = (100e3, 1e6, 3e6)


def core_path(z, band, Nhat, gate):
  """core.residual_mode with the gear's parameter set (product path)."""
  y, phi, st, dg = residual_mode(
      z, FS, BANDS[band]['fn'], Nhat, B_WIN,
      zeta=ZETA, tauG=TAU_G, Nt_win=NT_WIN, gate=gate, **gate_params(band))
  return y


def main():
  t0 = time.time()
  print('残差窗一致性验证 (审查项 #4): core.residual_mode vs '
        'validate_tracking.gear_filter')
  print(f'  fs={FS/1e6:.0f}MS/s, B_win={B_WIN/1e6:.0f}MHz, '
        f'NT_WIN={NT_WIN} taps, 容差 |Δerr| < {TOL_PP:.0f}pp')

  fails = []
  for tag, gate, cnr_db in (('near-noiseless / gate=always', 'always', None),
                            ('CNR=3dB noisy / gate=auto', 'auto', 3.0)):
    print(f'\n  [{tag}]')
    print(f"    {'gear':<7} {'f0':>7} | {'err gear_filter':>15} "
          f"{'err residual_mode':>17} | {'|diff| pp':>9}")
    for band in ORDER:
      for f0 in FREQS:
        sc = make_scene(f0)
        if gate == 'always':
          z, Nhat = clean_z(sc), 1e-10
        else:
          s2 = 10 ** (-cnr_db / 10)
          rng = np.random.default_rng(40_000 + int(f0 / 1e3))
          z = (np.exp(1j * sc['ph'])
               + complex_bandlimited_noise(N, FS, B_FRONTEND, s2, rng))
          Nhat = s2
        yg, _, _, _, _ = gear_filter(z, band, Nhat, gate=gate)
        yc = core_path(z, band, Nhat, gate)
        eg = amp_err_pct(vdisc(yg), sc)
        ec = amp_err_pct(vdisc(yc), sc)
        d = abs(eg - ec)
        ok = d < TOL_PP
        if not ok:
          fails.append((tag, band, f0, eg, ec, d))
        print(f"    {band:<7} {f0/1e3:5.0f}k | {eg:+14.3f}% {ec:+16.3f}% "
              f"| {d:9.4f}{'' if ok else '   <-- FAIL'}")

  print('\n' + '=' * 70)
  n_cases = 2 * len(ORDER) * len(FREQS)
  if fails:
    print(f'FAIL: {len(fails)}/{n_cases} 组合差异 >= {TOL_PP:.0f}pp')
  else:
    print(f'PASS: 全部 {n_cases} 组合 (三档 x 三频 x 无噪/含噪) '
          f'幅度误差差异 < {TOL_PP:.0f}pp -- 验证路径 = 产品路径')
  print(f'[elapsed {time.time()-t0:.1f} s]')
  return 1 if fails else 0


if __name__ == '__main__':
  raise SystemExit(main())
